### PR TITLE
Implement convenience factory methods for Async[Throwing]Stream

### PR DESCRIPTION
**Pull request by FranzBusch**
This is the implementation for https://github.com/apple/swift-evolution/pull/1824

### CODE REVIEW

File: test/Concurrency/Runtime/async_stream.swift
Comment (FranzBusch) [inline test/Concurrency/Runtime/async_stream.swift:12]: Reenabling the tests since the underlying problem with task cancellation was fixed by @ktoso

Comment (ktoso) [reply test/Concurrency/Runtime/async_stream.swift:12]: At least I certainly hope so! This was quite some time ago, thank you for enabling the tests again 💯

File: stdlib/public/Concurrency/AsyncStream.swift
Comment (ktoso) [inline stdlib/public/Concurrency/AsyncStream.swift:443]: formatting seems off, 2 spaces

**Review by ktoso — APPROVED** [review]
LGTM AFAICS! Pending evolution approval.

Need to decide what swift version this lands in and if we can get it into release/5.8 -  cc @beccadax or @DougGregor would know, but I guess it depends on quickness of the SE review

**Comment by FranzBusch** [comment]
@swift-ci please smoke test

**Comment by FranzBusch** [comment]
@swift-ci please smoke test

**Comment by beccadax** [comment]
@swift-ci build toolchain

**Comment by ktoso** [comment]
Seems the review is converging on using a tuple, might want to update this PR @FranzBusch  :)

**Comment by FranzBusch** [comment]
@swift-ci please smoke test

**Comment by ktoso** [comment]
Since a commit was pushed after the last trigger:

@swift-ci please smoke test

**Comment by ktoso** [comment]
ABI test failure, missing annotations on the following:

```

+Func AsyncStream.makeStream(of:bufferingPolicy:) is a new API without @available attribute
+Func AsyncThrowingStream.makeStream(of:throwing:bufferingPolicy:) is a new API without @available attribute
```

File: stdlib/public/Concurrency/AsyncStream.swift
Comment (ktoso) [inline stdlib/public/Concurrency/AsyncStream.swift:440]: This is missing

```suggestion
  @available(SwiftStdlib 5.1, *)
  @backDeployed(before: SwiftStdlib 5.9)
```

File: stdlib/public/Concurrency/AsyncThrowingStream.swift
Comment (ktoso) [inline stdlib/public/Concurrency/AsyncThrowingStream.swift:486]: This is missing

```suggestion
  @available(SwiftStdlib 5.1, *)
  @backDeployed(before: SwiftStdlib 5.9)
```

**Review by ktoso — APPROVED** [review]
I suggested how to fix the ABI checker issue inline; I believe that's the right combination but @tshortli could double check that.

File: stdlib/public/Concurrency/AsyncStream.swift
Comment (tshortli) [reply stdlib/public/Concurrency/AsyncStream.swift:440]: Since we removed the requirement that `@available` be present simultaneously with `@backDeployed` this is not technically necessary. `makeStream` will inherit its `introduced:` availability from the enclosing extension so the declaration will have appropriate availability. The suggested change here is unnecessary but could be preferred stylistically.

Comment (tshortli) [reply stdlib/public/Concurrency/AsyncStream.swift:440]: If the ABI checker is complaining, it may be being overly aggressive here.

Comment (tshortli) [reply stdlib/public/Concurrency/AsyncStream.swift:440]: Oh yeah I remember now. Yeah this is a limitation of the ABI checker. Annotating the individual APIs with availability as Konrad has suggested doesn't hurt and it does add clarity for readers so it's the right way to go.

Comment (ktoso) [reply stdlib/public/Concurrency/AsyncStream.swift:440]: Sounds good, thanks for conforming!

**Comment by FranzBusch** [comment]
@swift-ci please smoke test

**Comment by FranzBusch** [comment]
@swift-ci please smoke test

**Pull request merged**
